### PR TITLE
Suggest adding a lifetime constraint for opaque type

### LINKED ISSUE #67577: Non-sense diagnostic message from lifetime checker
I try to write method, that will return some iterator to filtered internal vector of strings
```rust
struct List {
  data: Vec<String>,
}
impl List {
  fn started_with<'a>(&'a self, prefix: &'a str) -> impl Iterator<Item=&'a str> {
    self.data.iter().filter(|s| s.starts_with(prefix)).map(|s| s.as_ref())
  }
}
```
But this code is not compiling with non-sense error, that recommends me to do thing, that already done:

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c7bd64a8b1c698ab819d8ddbc010486b))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0597]: `prefix` does not live long enough
 --> src/lib.rs:6:47
  |
5 |   fn started_with<'a>(&'a self, prefix: &'a str) -> impl Iterator<Item=&'a str> {
  |                   -- lifetime `'a` defined here     --------------------------- opaque type requires that `prefix` is borrowed for `'a`
6 |     self.data.iter().filter(|s| s.starts_with(prefix)).map(|s| s.as_ref())
  |                             ---               ^^^^^^ borrowed value does not live long enough
  |                             |
  |                             value captured here
7 |   }
  |   - `prefix` dropped here while still borrowed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.

```
Reproduced on all channels

### PR DESCRIPTION
**Pull request by ohadravid**
Fixes #67577, where code like this:

```
struct List {
  data: Vec<String>,
}
impl List {
  fn started_with<'a>(&'a self, prefix: &'a str) -> impl Iterator<Item=&'a str> {
    self.data.iter().filter(|s| s.starts_with(prefix)).map(|s| s.as_ref())
  }
}
```

will show this error:
```
   Compiling playground v0.0.1 (/playground)
error[E0597]: `prefix` does not live long enough
 --> src/lib.rs:6:47
  |
5 |   fn started_with<'a>(&'a self, prefix: &'a str) -> impl Iterator<Item=&'a str> {
  |                   -- lifetime `'a` defined here     --------------------------- opaque type requires that `prefix` is borrowed for `'a`
...
```

but without suggesting the lovely `help: you can add a constraint..`.

r? @estebank

### CODE REVIEW

File: src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs [hunk 1]
```diff
@@ -206,6 +206,28 @@ impl BorrowExplanation {
                         ),
                     );
                 };
+
+                if let ConstraintCategory::OpaqueType = category {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]: This code (including the `if let Constraint...`) could be in its own method.

File: src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs [hunk 2]
```diff
@@ -206,6 +206,28 @@ impl BorrowExplanation {
...
+                        let suggestable_name = if region_name.was_named() {
+                            region_name.to_string()
+                        } else {
+                            "'_".to_string()
+                        };
+
+                        err.span_suggestion(
+                            span,
+                            &format!(
+                                "you can add a constraint to the {}to make it last less than \
```
Comment (Centril) [inline src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]:
```suggestion
                                "you can add a bound to the {}to make it last less than \
```

Comment (ohadravid) [reply src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]: I used the same text as https://github.com/rust-lang/rust/blob/master/src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs#L57 (but I agree it's less "correct" than `bound`).
Should I change both or just this one?

Comment (estebank) [reply src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]: please change both

**Review by estebank — APPROVED** [review]
r=me after dealing with the nitpicks

File: src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs [hunk 3]
```diff
@@ -206,6 +206,47 @@ impl BorrowExplanation {
...
+                        region_name.to_string()
+                    } else {
+                        "'_".to_string()
+                    };
+
+                    err.span_suggestion(
+                        span,
+                        &format!(
+                            "you can add a constraint to the {}to make it last less than \
+                             `'static` and match {}",
```
Comment (estebank) [inline src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]:
```suggestion
                             `'static` and match `{}`",
```

**Comment by Mark-Simulacrum** [comment]
Looks like this might need to be rebased atop master (see https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16871/logs/1534 from https://github.com/rust-lang/rust/pull/67643#issuecomment-569142318).

@bors r-

File: src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs [hunk 4]
```diff
@@ -54,7 +54,7 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
                             err.span_suggestion(
                                 fn_return_span,
                                 &format!(
-                                    "you can add a constraint to the return type to make it last \
+                                    "you can add a bound to the return type to make it last \
                                  less than `'static` and match {}",
```
Comment (Mingun) [inline src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs]:
```
match {}
```
can be replaced with
```
match `{}`
```
to match already applied suggestion to `add_lifetime_bound_suggestion_to_diagnostic` method

File: src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs [hunk 5]
```diff
@@ -206,6 +206,47 @@ impl BorrowExplanation {
...
+                    err.span_suggestion(
+                        span,
+                        &format!(
+                            "you can add a bound to the {}to make it last less than \
+                             `'static` and match `{}`",
+                            category.description(),
+                            region_name,
+                        ),
+                        format!("{} + {}", snippet, suggestable_name),
+                        Applicability::Unspecified,
```
Comment (Mingun) [inline src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]: It seems that this fix can be applied automatically, isn't it?

Comment (ohadravid) [reply src/librustc_mir/borrow_check/diagnostics/explain_borrow.rs]: I used the same `Applicability` as the other suggestions for `impl Trait` lifetimes.
I don't think we can be **sure** this snippet will work (@estebank just fixed an issue with `;` for example)

**Comment by ohadravid** [comment]
@Mark-Simulacrum rebased and the CI is green 😄

**Comment by Mark-Simulacrum** [comment]
@bors r=estebank 

Thanks!

**Comment by ohadravid** [comment]
Ah it was `--compare-mode=nll`.. sorry about that! I `bless`ed it again and now it should also pass the full CI 😄 @Mark-Simulacrum

**Pull request merged**